askama/src/filters/humansize.rs

```diff
@@ -43,82 +41,190 @@ impl FastWritable for FilesizeFormatFilter {
+                .iter()
+                .take_while(|(_, m, _)| *m <= self.0)
+                .last()
+                .unwrap_or(&SI_PREFIXES[SI_PREFIXES.len() - 1]);
+            // divide number up into "full" units, and remainder
+            let integer = self.0 / unit.1;
+            let remainder = self.0 % unit.1;
+
+            // Compute fractional part with desired precision
+            // limit precision to the chosen unit's exponent - a greater precision will just
+            // generated tailing 0s that will be dropped.
+            let precision = self.1.min(unit.2);
+
+            // integer portion (before decimal point)
+            integer.write_into(dest, values)?;
+            if precision > 0 {
+                let mut scale = 10u128.pow(precision as u32);
+                let mut fractional = remainder / (unit.1 / scale);
+                if fractional > 0 {
+                    '.'.write_into(dest, values)?;
+                    for _ in 0..precision {
+                        scale /= 10;
+                        let digit = (b'0' + (fractional / scale) as u8) as char;
+                        digit.write_into(dest, values)?;
+                        fractional %= scale;
+                        if fractional == 0 {
+                            break;
+                        }
+                    }
+                }
+            }
+            ' '.write_into(dest, values)?;
+            unit.0.write_into(dest, values)?;
+            'B'.write_into(dest, values)?;
+
+            Ok(())
         }
     }
 }
 
-/// Formats `buffer` to contain the decimal point, decimal places and unit
-fn format_frac(buffer: &mut MaybeUninit<[AsciiChar; 8]>, prefix: AsciiChar, scaled: u32) -> &str {
-    // LLVM generates better byte code for register sized buffers
-    let buffer = buffer.write(AsciiStr::new_sized("..0 kB"));
-    buffer[4] = prefix;
-
-    let frac = scaled % 100;
-    let buffer = if frac == 0 {
-        &buffer[3..6]
-    } else {
-        let digits = AsciiChar::two_digits(frac);
-        if digits[1] == AsciiChar::new(b'0') {
-            // the decimal separator '.' is already contained in buffer[1]
-            buffer[2] = digits[0];
-            &buffer[1..6]
-        } else {
-            // the decimal separator '.' is already contained in buffer[0]
-            [buffer[1], buffer[2]] = digits;
-            &buffer[0..6]
-        }
-    };
-    AsciiStr::from_slice(buffer).as_str()
-}
+/// (`si_prefix`-char, `amount of bytes corresponding to 1 unit`)
+const SI_PREFIXES: &[(char, u128, u8)] = &[
+    ('k', 1_000, 3),
+    ('M', 1_000_000, 6),
+    ('G', 1_000_000_000, 9),
+    ('T', 1_000_000_000_000, 12),
+    ('P', 1_000_000_000_000_000, 15),
+    ('E', 1_000_000_000_000_000_000, 18),
+    ('Z', 1_000_000_000_000_000_000_000, 21),
+    ('Y', 1_000_000_000_000_000_000_000_000, 24),
+    ('R', 1_000_000_000_000_000_000_000_000_000, 27),
+    ('Q', 1_000_000_000_000_000_000_000_000_000_000, 30),
+];
```

Comment:
Good idea to store the maximum number of trailing digits as part of the table! Maybe use a const fn to make it a bit more readable?
```rust
const SI_PREFIXES: &[(char, u128, u8)] = &[
    si_prefix('k', 3),
    si_prefix('M', 6),
    si_prefix('G', 9),
    si_prefix('T', 12),
    si_prefix('P', 15),
    si_prefix('E', 18),
    si_prefix('Z', 21),
    si_prefix('Y', 24),
    si_prefix('R', 27),
    si_prefix('Q', 30),
];

const fn si_prefix(prefix: char, exponent: u8) -> (char, u128, u8) {
    (prefix, 10u128.pow(exponent as u32), exponent)
}
```